add FCStd read and write nodes

rastart opened the pull request:
## new nodes (exchange)

FCStd read - import parts from single or multiple freecad project files
filter by name feature 

FCStd read  Sketches - import Sketches from single or multiple freecad project files
filter by Sketches name

FCStd write - write solid or mesh parts in a single freecad project file

FCStd Spreadsheet - read/write parameter from a freecad spreadsheet

Approximate Subd to Nurbs - convert an obj (with subd modifier) to NURBS

## FCStd read features

**filter by name  -dropdown enum list**
select the part to be imported

**read_update  -bool**
enable/disable reading

**read_body  -bool**
enable/disable reading freecad Body

**read_body  -bool**
enable/disable reading freecad Parts

**tool_parts -bool**
enable/disable reading freecad parts/features inside a solid operation or body ( it read just the final output)

**invert filter -bool**
invert filter selection

**input-> file path -path**
**input-> part list -string** (OPTIONAL: part labes in string list format)
**output-> -solid**

## FCStd Sketch read features

**mode  -dropdown enum list**
read mode:  only geometry , only construction geometry, both

**filter by name  -dropdown enum list**
select the sketch to be imported

**max_points  -int**
parameter to resample the geometry by set max points in relation the the max geometry curve length

**invert filter -bool**
invert filter selection

**read_update  -bool**
enable/disable reading

**input-> file path -path**
**input-> part list -string** (OPTIONAL: part labes in string list format)
**output-> -verts**
**output-> -edges**
**outputs-> -curve**

## FCStd write features
**write_name  -string**
name of the part(s) to be written
a suffix  `_SV_` and a numeration will be added to the name 
to avoid delete erroneusly parts in the freecad project.
The writed parts will be overwrited on node update.

**format  -enum**
select mesh or solid format

**write_update  -bool**
enable/disable writing

**input-> file path -path**

solid mode:
**input-> -solid**
mesh mode:
**input-> -verts**
**input-> -faces**

## FCStd Spreadsheet

**filter Spreadsheets by name  -dropdown enum list**
select the freecad Spreadsheet

**filter Spreadsheets cell by alias  -dropdown enum list**
select the freecad cell alias in the selected Spreadsheet

**write parameter**
set the node to write mode

**input-> file path -path**
**input-> cell in -float**
**output-> cell out -float**

## important
Code tested both in blender 2.93 (py3.9) and freecad 0.2 (py3.9)
More tests required to ensure safe usability.

- [x] FCStd read node: needs label list input to filter parts
- [x] FCStd Sketch read node: need testing, complete curve translation and a better label...
- [x] FCStd read node: write dropdown freecadpart list node 
- [x] FCStd spreadsheet node: read / write mode
- [x] code refactoring
- [x] fix for blender 2.93 / freecad 0.20 py3.9
- [x] add an update button
- [x] Approximate Subd to Nurbs node added
- [ ] Approximate Subd to Nurbs node sorting algorithm improving
- [ ] Approximate Subd to Nurbs node debugging
- [ ] nodes review / testing
- [ ] add docs
- [ ] ready to merge

![update_buttons](https://user-images.githubusercontent.com/8501792/119225042-c4707d80-bb01-11eb-99aa-16da210bf461.jpg)

vicdoval commented:
Some docs would be nice before merging... :)

rastart commented:
ok, are there some example I can follow?

here some screen about use cases:

### read node without filter keys, import all freecad parts visible/hidden

![read_all](https://user-images.githubusercontent.com/8501792/97240032-71de1580-17ed-11eb-95c3-89b6e5511ed2.jpg)


### read node with filter keys

![read_filter](https://user-images.githubusercontent.com/8501792/97240034-7276ac00-17ed-11eb-94ae-88f5781a6cd9.jpg)

### read node with filter keys 
**for part design is important to filter the final step of the body tree**

![read_pt_design](https://user-images.githubusercontent.com/8501792/97240841-a652d100-17ef-11eb-808d-374a2be0714c.jpg)

### write node - solid mode

![write_solid](https://user-images.githubusercontent.com/8501792/97240038-730f4280-17ed-11eb-9e23-76df95109b06.jpg)


### write node - mesh mode
![write_mesh](https://user-images.githubusercontent.com/8501792/97240036-730f4280-17ed-11eb-983c-5e92d2f43761.jpg)

Durman commented on nodes/exchange/FCStd_read.py:
```diff
@@ -0,0 +1,103 @@
+    if ',' in part_filter:
+        part_filter = part_filter.split(',')
+
+    elif len(part_filter)==0:
```
```suggestion
    elif not part_filter:
```

Durman commented on nodes/exchange/FCStd_read.py:
```diff
@@ -0,0 +1,103 @@
+def LoadSolid(fc_file,part_filter,inv_filter):
+    solids = []
+
+    if ',' in part_filter:
+        part_filter = part_filter.split(',')
+
+    elif len(part_filter)==0:
+        part_filter = ['all']
```
It is not too good to use one variable for two different things. I mean `part_filter` should be either string ot list.

rastart replied:
Yes, I'm trying another solution adding a filter input . 
I didn't find any string input node to use so I scripted a simple one to make a test.
It works better and it's a lot more consistent also if require a lot of nodes if one need to import a lot of parts, indeed it could be added in the future a specific node to read parts label and pass the wanted selection.

![filter_input](https://user-images.githubusercontent.com/8501792/97336074-e319d900-187e-11eb-943b-be1b1d8789d9.png)

Durman commented on nodes/exchange/FCStd_read.py:
```diff
@@ -0,0 +1,103 @@
+                    if not obj.Label in part_filter:
+                        solids.append(obj.Shape)
+
+        F.closeDocument(Fname)
+
+    except:
+        print ('FCStd read error')
+        return []
+
```
```suggestion
    except:
        info('FCStd read error')
    finally:
        F.closeDocument(Fname)
```

Durman commented on nodes/exchange/FCStd_write.py:
```diff
@@ -0,0 +1,182 @@
+                return             
+            
+
+def FC_WRITE_PARTS(fc_file,verts,faces,part_name,solid,mod):
```
```suggestion
def fc_write_parts(fc_file, verts, faces, part_name, solid, mod):
```

vicdoval commented on nodes/exchange/FCStd_read.py:
```diff
@@ -0,0 +1,102 @@
+    from numpy import ndarray
+
+    class SvReadFCStdNode(bpy.types.Node, SverchCustomTreeNode):
+        ''' SvReadFCStdNode '''
```
We have a new convention for the nodes docstrigns:

https://github.com/nortikin/sverchok/blob/a95871e3a5a77b13e63fda4654a07a638f6ab19d/nodes/field/attractor_field_mk2.py#L28-L31
The triggers appear on the search menu. The tooltip when you place the mouse over the node name in the menu

vicdoval commented on nodes/exchange/FCStd_read.py:
```diff
@@ -0,0 +1,102 @@
+        
+        read_update : BoolProperty(name="read_update", default=True)
+        part_filter : StringProperty(name="part_filter", default="", description="use ',' to separate name with no space: part1,part2,... ")
+        inv_filter : BoolProperty(name="inv_filter", default=False)
```
If you want the node to update on property change you would need to get the `updateNode `function from data_structure as here:
https://github.com/nortikin/sverchok/blob/a95871e3a5a77b13e63fda4654a07a638f6ab19d/nodes/curve/apply_field_to_curve.py#L8
and add a update keyword to the property calling to that function
https://github.com/nortikin/sverchok/blob/a95871e3a5a77b13e63fda4654a07a638f6ab19d/nodes/curve/apply_field_to_curve.py#L30-L33

vicdoval commented on nodes/exchange/FCStd_write.py:
```diff
@@ -0,0 +1,182 @@
+    from bpy.props import IntProperty, FloatProperty, StringProperty, BoolProperty,EnumProperty
+    from sverchok.node_tree import SverchCustomTreeNode, throttled
+    from sverchok.data_structure import updateNode
+    from numpy import ndarray
```
I think you are not using  `sys`, `IntProperty`, `FloatProperty` and `ndarray`

vicdoval commented on nodes/exchange/FCStd_write.py:
```diff
@@ -0,0 +1,182 @@
+
+            else:
+                self.inputs.new('SvSolidSocket', 'Solid')
+           
```
When this function is called obj_format will always be the default value (solid)

rastart replied:
Yes,  I noticed a strange behaviour with the enum, also if I save somethime when I reopen the file the value is changed. 
I'll look better at how this is implemented in other nodes.

vicdoval commented on nodes/exchange/FCStd_write.py:
```diff
@@ -0,0 +1,182 @@
+
+                if any((self.inputs['Verts'].is_linked,self.inputs['Faces'].is_linked,self.write_update)):
+
+                    verts=self.inputs['Verts'].sv_get()
+                    pols=self.inputs['Faces'].sv_get()
+                    fc_write_parts(fc_file,verts,pols,self.part_name,None,self.obj_format)
```
`verts_in = self.inputs['Verts'].sv_get(deepcopy=False) ` 
`pols_in = self.inputs['Faces'].sv_get(deepcopy=False)` because you don't modify data in place this makes the sv_get faster
`verts, pols= match_long_repeat([verts_in, pols_in])` to match different list lenghts (`match_long_repeat `has to be imported from data_structure)
`fc_write_parts(fc_file, verts, pols, self.part_name, None, self.obj_format)`
please try to keep PEP8 style (space after coma...)

rastart commented:
FCSdtnode updated with dropdown part list selection

 <img src= "https://user-images.githubusercontent.com/8501792/97753160-9986f380-1af5-11eb-9d2d-2711869f1176.png" width="500" height="auto">

rastart commented:
added read sketch node!
The sketch geometry is converted in polylines by freecad 'discretize' function and passed to verts/edge data.
Added a resample function with a parameter 'max_points' that set the max number of point in relation to the max length in the sketch geometry.

![fcnodes_sketch](https://user-images.githubusercontent.com/8501792/100556356-3af29800-32a2-11eb-9480-d9ef13b7d094.png)

JuanJ67m commented:
share samples please, and create nodes for LibreOffice or excel please.

vicdoval commented on nodes/exchange/FCStd_sketch.py:
```diff
@@ -0,0 +1,285 @@
+            self.inputs.new('SvStringsSocket', "Part Filter")   
+
+            self.outputs.new('SvVerticesSocket', "Verts")
+            self.outputs.new('SvVerticesSocket', "Edges")
```
If the edges socket outputs indices should be a `SvStringsSocket`

rastart commented:
> share samples please, and create nodes for LibreOffice or excel please.

I ll work on some samples, for generic Spreadsheet functionality you can see here: https://github.com/nortikin/sverchok/issues/3639

rastart commented:
sketch node now can output all kind of freecad sketch geometry
added option to filter construction geometry

![fc_sketch_node](https://user-images.githubusercontent.com/8501792/106362303-56216880-6322-11eb-91e2-b00864032aa4.jpg)

nortikin commented:
have not such experience in FC. maybe someone can test it and merge at last.

rastart commented:
> have not such experience in FC. maybe someone can test it and merge at last.

I started to make some little projects with exchange freecad nodes and It's going almost smooth, 
it make me possible to easily solve design issues ( obviuosly in hybrid situation: mesh + solid ) more than I could do in other commercial solution, with the develpment of freecad ( now in 0.19) this could be more and more interesting.
At this point I have to make some examples file ( .blend + .fcstd ) to share, also in freecad forum. Let's see...

vicdoval commented:
@rastart do you have a freecad 0.19 python 3.9 version? (for blender 2.92) 😄

rastart commented:
> @rastart do you have a freecad 0.19 python 3.9 version? (for blender 2.92) 😄

Still using 0.18 py37 on Blender 2.91, I checked freecad 0.19.1 and it's packed with py 3.8...
What a pain....

rastart commented:
> @rastart do you have a freecad 0.19 python 3.9 version? (for blender 2.92) 😄

I confirm last "FreeCAD_weekly-builds-24612-Win-Conda_vc14.x-x86_64" is packed with python 3.9.2!!!!
I'll try it in the next days

nortikin commented:
https://vk.com/wall-35076122_12458 news news news

nortikin commented:
exchange index with Gcode node added

rastart commented:
> exchange index with Gcode node added
Interesting... I'm looking forward to try this node on my printer!!!

fcstd nodes:
I just fixed some issues with sketch costrution mode reading from sketch read node (freecad 0.3 py3.,9 release related), I haven't noticed any other issues.
At this point I think I would add a button to update the nodes manually and keep an autoupdate option.

rastart commented:
added an update button:
![update_buttons](https://user-images.githubusercontent.com/8501792/119224207-a99c0a00-bafd-11eb-80fb-509c544019bb.jpg)

I'm already making some little projects with these nodes and I'll make a new repo on github just for project example files (freecad/blender) 
the nodes have to be tested and debugged in a freecad/blender workflow, so to improve them It will be required to collect some use cases, once I'll have something I'll share it.

portnov commented:
Hmm
The last node you added is very promising :)
![Screenshot_20210817_233226](https://user-images.githubusercontent.com/284644/129781359-aca0a665-7301-41d7-9d2a-1978f02d4d29.png)
I do not understand how, but it works somehow :)

Although it appears it has some problems with normals...

and it could have better handling of situation when the object does not have modifiers.

rastart commented:
There is something strange in  solid viewer but the result solid is good, I made also some test in freecad.
![jar_sv_fcad_workflow](https://user-images.githubusercontent.com/8501792/129783347-1d593906-3d3a-4720-a578-345ab4715de2.png)
But yes, it works someway...

The approach is this:
1 get the quad edge border
2 subdivide 
3 recall the original border and get the center vertex
4 use the border to build a surface patch
5 use the center vertex of the subdivided quads to constraint the surface

patch = F.ActiveDocument.addObject("Surface::Filling","Surface%s"%item) # 4
patch.BoundaryEdges = (com_obj, edge_names) # 4 
patch.Points = (CEN, "Vertex1") # 5 THE MAGIC 
Before I spotted this point constraints feature in freecad I couldn't get a good translation of the subd patch.

The main problem now is the sorting algorithm I used for the step n3: too many FOR cycles and bottleneck for complex mesh.

HaasNL2 commented:
This is such good stuff, I hate to see it go to waste because people can't find it or already give up on installing. So please merge it!

luzpaz commented:
Thanks for all the effort to get this working thus far.  
Question: what is blocking this branch from being merged?

nortikin commented:
I want to merge this. How do you do?

portnov commented:
Eeh. CI is broken :/

luzpaz commented:
![Screenshot_20220412_121952](https://user-images.githubusercontent.com/4140247/163008455-9e6df93c-a572-42d3-8b9f-adcba61fafea.png)
